text/0000-inferred-variant-constructor.md

```diff
@@ -0,0 +1,204 @@
+- Feature Name: `inferred_variant_constructor`
+- Start Date: 2017-03-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow the name (qualifier) of an enum variant to be elided in expressions and
+patterns whenever it can be inferred.
+
+# Motivation
+[motivation]: #motivation
+
+The main goal of this proposal is to improve ergonomics of enums in Rust by
+reducing verbosity.  Currently, there are two convenient ways to use enums.
+One would either:
+
+  - bring the enum type into scope (`use foolib::Foo;`), or
+  - bring the enum variants into scope (`use foolib::Foo::*;`).
+
+The former is less convenient than the latter, but the latter pollutes the
+namespace more.  One could always write enum variants fully qualified,
+`foolib::Foo::Alpha`, but that is even more tedious.
+
+This proposal intends to make variant expressions as convenient and
+pollution-free as field expressions today.  Whereas field expressions allow
+the user to extract fields of a struct without ever bringing the struct into
+scope, inferred variant constructors allow the user to construct/match
+variants of an enum without ever bringing the enum or its variants into scope.
+
+This could lower the usability barrier for enums, particularly for library
+users.  It would reduce the verbosity of enums to a level similar to
+dynamically typed languages, where strings (e.g. Python) or symbols
+(e.g. Lisps, Ruby) designate ad hoc variants
+
+The feature could encourage library writers to use domain-specific enums more
+often, rather than to re-purpose generic enums like `Option` or `Result`,
+leading to more readable code.  It would also allow library writers to use
```

Comment:
> Option and Result have lots of functions on them, while custom enums don't, or first need them to be defined. They are very useful and allow for patterns like chaining.

They can be, but they are not always needed.  The proposal is focused on reducing the boilerplate for “one-off” uses of enums, with a particular emphasis on users of multiple different libraries.

> And library authors can just declare a prelude with all of their enum variants inside. Or am I mistaken? idk...

Preludes compound the risk of naming collisions between different libraries due to its naming pollution.    They are also risky for library authors as any addition of items to a prelude can cause breakage downstream.  It would also be difficult for preludes to handle multiple enums in which some of the variants have the same name.